File: 06-Greedy Approach/Fractional Knapsack code.py
def knapsack(p,w,n,weight):
       cost=0
       for i in range(0,n):
              if w[i]>weight:
                     cost+=(weight/w[i])*p[i]
                     break;
              else:
                     weight-=w[i]
                     cost+=p[i]
       return cost

n=int(input("Enter the number of objects "))
weight=int(input("Enter the weight a knapsack can carry "))
p=[]
w=[]
ratio=[]
for i in range(0, n):
       p.append(int(input("profit of object")))
       w.append(int(input("weight of object ")))
       ratio.append(p[i]/w[i])

print(p,w,ratio)

# sorting of the ratio in decreasing order
for i in range(0,n-1):
       for j in range (i+1,n):
              if ratio[i]<ratio[j]:
                     ratio[i],ratio[j]=ratio[j],ratio[i]
                     w[i],w[j]=w[j],w[i]
                     p[i], p[j]=p[j],p[i]
# Now according to the greedy approach, we need to include the objects which have maximum profit by weight ratio
print(p,w,ratio)
print("The maximum profit is ",round(knapsack(p,w,n,weight),2))

File: 06-Greedy Approach/Fractional Knapsack code.cpp
#include<iostream>
using namespace std;

float knapsack(float p[],float w[], float weight,int n)
{
	float cost=0;
	int i=0;
	for(i=n-1;i>=0;i--)
	{
		if(w[i]>weight)
		{
			break;
		}
		cost=cost+p[i];
		weight=weight-w[i];
	}
	if(i>=0)
	{
		cost=cost+(weight/w[i])*p[i];
	}
	return cost;
}
int main()
{
	int n;
	float weight;
	cout<<"Enter the number of objects"<<endl;
	cin>>n;
	cout<<"Ente the amount the knapsack can contain"<<endl;
	cin>>weight;
	float p[n],w[n],r[n];
	cout<<"Enter the profit and weight of the object"<<endl;
	for(int i=0;i<n;i++)
	{
		cout<<"Enter the profit "<<" ";
		cin>>p[i];
		cout<<"Enter the weight "<<" ";
		cin>>w[i];
		r[i]=p[i]/w[i];
		cout<<endl;
	}
	// now to sort the ratio array and according to that sorting the weight and profit array 
	for(int i=0;i<n-1;i++)
	{
		for(int j=i+1;j<n;j++)
		{
			if( r[i]>r[j])
			{
				float temp=r[i];
				r[i]=r[j];
				r[j]=temp;
				temp=p[i];
				p[i]=p[j];
				p[j]=temp;
				temp=w[i];
				w[i]=w[j];
				w[j]=temp;
			}
		}
	}
	cout<<"The sorted arrays are "<<endl;
	for(int i=0;i<n;i++)
	{
		cout<<"profit "<<p[i]<<" weight "<<w[i]<<" ratio "<<r[i]<<endl;
	}
	cout<<"The maximum profit would be "<< knapsack(p,w,weight,n);
	// Now to maximise the profit we need to calculate the profit/weight 
	return 0;	
}
